### ISSUE TITLE
Issue with Monte Carlo barostat, constraints, and PBC

### ISSUE THREAD

**Issue by mark-cresset** [openmm/openmm#3377]
Hi all,

I might be doing something wrong here, but I'd appreciate some advice. I'm running some unit tests on our protein equilibration pathway, and have noticed a dependency on how the system is set up. The workflow that demonstrates the problem is:

1) Take the input structure and minimise it. 
2) Add a constraint force for backbone protein atoms holding them to remain near where the coordinates at the end of step (1) are
3) Run a short NVT to get the temperature up
4) Run a short NPT to get the pressure right.

I notice a different behaviour depending on the coordinates of the protein in the periodic box. If the protein is in the middle of the box, then this works fine. If the protein is instead centred on the origin, then the barostat really struggles to equilibrate the pressure in step (4). If the constraint force is not present then the barostat is fine.

The system in question has several disconnected chains, so my current guess as to what is happening is that at the end of step (1) when the protein is centred on the origin at least one chain gets wrapped (my understanding is that wrapping occurs individually for connected systems). We then nail the atoms into place in step 2, and as a result the barostat has real problems as when it tries to shrink the cell the wrapped chain gets dragged away from its constrained position and the corresponding energy increase causes MC fails. Is this plausible?

We can work around this by ensuring that the system starts up with the protein centred in the box (which is the default if we set the system up with the Amber tools), at which point it's extremely unlikely that any components are going to move far enough to be wrapped in the short equilibration. However, I was wondering whether there was an alternative way to solve the issue: could I e.g. add a dummy bond between the protein components to ensure that they all get treated as one unit rather than individually by the barostat?

If some example code would be useful then happy to provide: I'd just need a bit of time to extract a simple test case out of our existing code which has all sorts of irrelevant complications in it...

**Comment by peastman** [comment]
How do you implement the restraint force?  Assuming it's with CustomExternalForce, are you using the `periodicdistance()` function as discussed in the [API docs](http://docs.openmm.org/latest/api-c++/generated/CustomExternalForce.html)?

**Comment by mark-cresset** [comment]
The restraint force uses the periodicdistance function:

force = openmm.CustomExternalForce("k*periodicdistance(x,y,z,x0,y0,z0)^2")

**Comment by mark-cresset** [comment]
I've created a small test example showing the problem. 
[barostat_problem.tar.gz](https://github.com/openmm/openmm/files/7711113/barostat_problem.tar.gz)

**Comment by peastman** [comment]
Your two sets of files are quite different from each other.  The OK ones have 40,302 atoms while the NOTOK ones have 47,592 atoms.  It isn't just a question of whether the protein is centered.

Can you reproduce the problem when using exactly the same system in both cases, and the only difference is the initial coordinates?

**Comment by mark-cresset** [comment]
Attached is a copy of the Input.inpcrd.NOTOK coordinates where I have just added <30.0,30.0,30.0> to each position. With these shifted coordinates the barostat equilibrates to a density of 1 after around 18K steps, while with the unshifted coordinates the density is still at 0.95 after 25K steps and is increasing very slowly.
[Input.inpcrd.NOTOK.shifted.gz](https://github.com/openmm/openmm/files/7734958/Input.inpcrd.NOTOK.shifted.gz)

**Comment by mark-cresset** [comment]
I'm hoping to be able to re-test this soon: should I expect that the fix for #3260 in 7.7 fix this?

**Comment by peastman** [comment]
I don't think so.  I was able to reproduce the behavior you saw, but haven't had time to try to diagnose the cause yet.
